src/term/transform/flatten.rs

```diff
@@ -207,17 +204,19 @@ fn make_split_rule(old_rule: &Rule, other_rule: &Rule, def_names: &DefNames) ->
         let new_ctr = Term::Tup { fst: Box::new(fst_arg), snd: Box::new(snd_arg) };
         new_body.subst(other_arg, &new_ctr);
       }
-      (Pattern::Var(..), _) => {
-        new_pats.push(other_arg.clone());
-      }
+      (Pattern::Var(..), _) => new_pats.push(other_arg.clone()),
       // Unreachable cases, we only call this function if we know the two patterns match together
       (
-        Pattern::Ctr(..) | Pattern::Num(..) | Pattern::Tup(..),
+        Pattern::Ctr(..) | Pattern::Tup(..),
         Pattern::Ctr(..) | Pattern::Num(..) | Pattern::Tup(..),
       ) => {
         unreachable!()
       }
-      (_, Pattern::Var(None)) => unreachable!(),
+      (Pattern::Ctr(..), Pattern::Var(None)) => new_pats.push(Pattern::Var(None)),
```

Comment:
Shouldn't you push one * pattern for each field that is being extracted from the nested pattern?
Try testing this:
```
data Tuple = (Pair a b)
(Foo (Pair (Pair a b) c)) = ...
(Foo *) = ...
```